test/js/web/fetch/body-cloning.test.ts

```diff
@@ -0,0 +1,737 @@
+      }
+    });
+    await requestServer({
+      method: "POST",
+      body: data,
+      headers: {
+        "Content-type": "application/octet-stream",
+      },
+    });
+  });
+
+  it("can clone all default types", async done => {
+    const data = JSON.stringify({ message: "Hello world" });
+    await startServer(async request => {
+      try {
+        const text_clone = request.clone();
+        const arrayBuffer = text_clone.clone();
+        const [array_buf, text, json] = await Promise.all([
+          arrayBuffer.arrayBuffer(),
+          text_clone.text(),
+          request.json(),
+        ]);
+        expect(array_buf).toEqual(Buffer.from(data).buffer.slice());
+        expect(text).toEqual(data);
+        expect(json).toEqual(JSON.parse(data));
+        done();
+        return new Response(JSON.stringify({}), {
+          headers: {
+            "Content-type": "application/json",
+          },
+        });
+      } catch (err) {
+        done(err);
+      }
+    });
+    await requestServer({
+      method: "POST",
+      body: data,
+      headers: {
+        "Content-type": "application/json",
+      },
+    });
+  });
+
+  it("can clone slow connection", async done => {
+    const data = JSON.stringify({ message: "Hello world" });
+    let socket;
+    await startServer(async request => {
+      try {
+        const text_clone = request.clone();
+        const arrayBuffer = text_clone.clone();
+        const readerClone = arrayBuffer.clone();
+        const [array_buf, text, json, reader_out] = await Promise.all([
+          arrayBuffer.arrayBuffer(),
+          text_clone.text(),
+          request.json(),
+          Bun.readableStreamToArrayBuffer(readerClone.body),
+        ]);
+        expect(array_buf).toEqual(Buffer.from(data).buffer.slice());
+        expect(reader_out).toEqual(Buffer.from(data).buffer.slice());
+        expect(text).toEqual(data);
+        expect(json).toEqual(JSON.parse(data));
+        socket.end();
+        done();
+        return new Response(JSON.stringify({}), {
+          headers: {
+            "Content-type": "application/json",
+          },
+        });
+      } catch (err) {
+        done(err);
+      }
+    });
+    socket = await Bun.connect({
+      hostname: server.hostname,
+      port: server.port,
+
+      socket: {
+        data(socket, data) {},
+        error(socket, error) {
+          done(error);
+        },
+        connectError(socket, error) {
+          done(error);
+        },
+      },
+    });
+    socket.write(
+      [
+        "POST / HTTP/1.1",
+        "Host: localhost",
+        "User-agent: test-client",
+        "Connection: keep-alive",
+        "Content-type: application/json",
+        "Content-length: " + data.length,
+        "\r\n",
+      ].join("\r\n"),
+    );
+    await sleep(500);
+    for (const char of data.split("")) {
+      socket.write(`${char}`);
+      await sleep(150);
+    }
+    socket.write("\r\n\r\n");
+    socket.flush();
```

Comment:
this doesn't check that the contents were actually written

Reply:
@Jarred-Sumner why would you need to check that if the test is performed on the server

Reply:
if the data is received on the server it HAS to have been written

Reply:
`socket.write` returns the number of bytes written. `socket.flush` does too. if the number of bytes written is not the expected number, the test will be flaky (for reasons unrelated to whether or not the code being tested works)